File: peer-review/TTK4145-2025-code-peer-review-files/3cee6c8f/Project/elev_algo_go/fsm/fsm.go
package fsm

import (
	"Driver-go/elevio"
	"fmt"

	"github.com/anonym/TTK4145-Project/Project/elev_algo_go/elevator"
	"github.com/anonym/TTK4145-Project/Project/elev_algo_go/requests"
	"github.com/anonym/TTK4145-Project/Project/elev_algo_go/timer"
)

var elev elevator.Elevator

func Init() {
	elev.Initialize()
}

func setAllLights(elev elevator.Elevator) {
	for floor := 0; floor < elevator.NumFloors; floor++ {
		for btn := 0; btn < elevator.NumButtons; btn++ {
			elevio.SetButtonLamp(elevio.ButtonType(btn), floor, elev.Requests[floor][btn])
		}
	}
}

func OnInitBetweenFloors() {
	fmt.Println("FSM: On init between floors")
	elevio.SetMotorDirection(elevio.MD_Down)
	elev.Dirn = elevio.MD_Down
	elev.Behaviour = elevator.EB_Moving
}

func OnRequestButtonPress(btn_floor int, btn_type elevio.ButtonType) {
	fmt.Println("FSM: On request button press")
	switch elev.Behaviour {
	case elevator.EB_DoorOpen:
		if requests.ShouldClearImmediately(elev, btn_floor, btn_type) {
			timer.Start(float64(elev.Config.DoorOpenDuration_s))
		} else {
			elev.Requests[btn_floor][btn_type] = true
		}
	case elevator.EB_Moving:
		elev.Requests[btn_floor][btn_type] = true
	case elevator.EB_Idle:
		elev.Requests[btn_floor][btn_type] = true

		// Additional logic to handle state transition from Idle
		pair := requests.ChooseDirection(elev)
		elev.Dirn = pair.Dirn
		elev.Behaviour = pair.Behaviour
		switch pair.Behaviour {
		case elevator.EB_DoorOpen:
			elevio.SetDoorOpenLamp(true)
			timer.Start(elev.Config.DoorOpenDuration_s)
			elev = requests.ClearAtCurrentFloor(elev)
		case elevator.EB_Moving:
			elevio.SetMotorDirection(elevio.MotorDirection(elev.Dirn))
		case elevator.EB_Idle:
			// Do nothing
		}
	}
	// Update lights
	setAllLights(elev)
	fmt.Println("\nNew state:\n")
	elev.Elevator_print()
}

func OnFloorArrival(newFloor int) {
	fmt.Println("FSM: On floor arrival")
	fmt.Println("Floor arrived at: ", newFloor)
	elev.Elevator_print()

	elev.Floor = newFloor
	elevio.SetFloorIndicator(elev.Floor)

	switch elev.Behaviour {
	case elevator.EB_Moving:
		if requests.ShouldStop(elev) {
			elevio.SetMotorDirection(elevio.MD_Stop)
			elevio.SetDoorOpenLamp(true)
			elev = requests.ClearAtCurrentFloor(elev)
			timer.Start(elev.Config.DoorOpenDuration_s)
			setAllLights(elev)
			elev.Behaviour = elevator.EB_DoorOpen
		}
	default:
		// Do nothing
	}
	fmt.Println("New state: ")
	elev.Elevator_print()
}

func OnDoorTimeout() {
	fmt.Println("FSM: Door timeout")
	elev.Elevator_print()

	switch elev.Behaviour {
	case elevator.EB_DoorOpen:
		pair := requests.ChooseDirection(elev)
		elev.Dirn = pair.Dirn
		elev.Behaviour = pair.Behaviour
		fmt.Println("Timeout before second case")
		fmt.Println("New direction: ", elevator.DirnToString(elev.Dirn))
		fmt.Println("New behaviour: ", elevator.EBToString(elev.Behaviour))
		
		switch elev.Behaviour {
		case elevator.EB_DoorOpen:
			timer.Start(elev.Config.DoorOpenDuration_s)
			elev = requests.ClearAtCurrentFloor(elev)
			setAllLights(elev)
			break
		case elevator.EB_Moving:
			elevio.SetDoorOpenLamp(false)
			elevio.SetMotorDirection(elevio.MotorDirection(elev.Dirn))
		case elevator.EB_Idle:
			elevio.SetDoorOpenLamp(false)
			elevio.SetMotorDirection(elevio.MotorDirection(elev.Dirn))
			break
		}
	}
	fmt.Println("New state: ")
	elev.Elevator_print()
}

// Something YMSE, IDK
func OnObstruction(obstr bool) {
	fmt.Println("FSM: On obstruction")
	elev.Elevator_print()

	switch elev.Behaviour {
	case elevator.EB_DoorOpen:
		if obstr {
			timer.Stop()
		} else {
			timer.Start(elev.Config.DoorOpenDuration_s)
		}
	}
	fmt.Println("New state: ")
	elev.Elevator_print()
}

File: peer-review/TTK4145-2025-code-peer-review-files/975af7d1/snapshot/code/src/single_elevator/fsm.rs
use driver_rust::elevio::elev as e;

use crate::config::NUM_BUTTONS;
use crate::single_elevator::elevator;
use crate::single_elevator::requests;
use crate::types::Direction;

pub fn set_all_lights(elevio_driver: &e::Elevator, elevator_state: &elevator::State) {
    for f in 0..elevator_state.config.number_of_floors {
        for c in 0..NUM_BUTTONS {
            let light = elevator_state.get_request(f.try_into().unwrap(), c.try_into().unwrap());
            elevio_driver.call_button_light(f.try_into().unwrap(), c, light);
        }
    }
}

pub fn on_init_between_floors(elevio_driver: &e::Elevator, elevator_state: &mut elevator::State) {
    elevio_driver.motor_direction(e::DIRN_DOWN);
    elevator_state.direction = Direction::Down;
    elevator_state.behaviour = elevator::Behaviour::Moving;
}

pub fn on_new_order_assignment(
    // This is almost a blind copy of fsm::on_request_button_press
    elevio_driver: &e::Elevator,
    mut elevator_state: &mut elevator::State,
    // request_floor: u8,
    // button: elevator::Button,
) {
    match elevator_state.behaviour {
        elevator::Behaviour::Idle => {
            // elevator_state.set_request(request_floor, button, true);
            let direction_behaviour_pair = requests::choose_direction(&elevator_state);
            elevator_state.direction = direction_behaviour_pair.0;
            elevator_state.behaviour = direction_behaviour_pair.1;
            match elevator_state.behaviour {
                elevator::Behaviour::Idle => {}
                elevator::Behaviour::DoorOpen => {
                    elevio_driver.door_light(true);
                    elevator_state.start_door_timer();
                    elevator_state = requests::clear_at_current_floor(elevator_state);
                    // THIS NEEDS TO BE BROADCAST AS WELL
                }
                elevator::Behaviour::Moving => {
                    elevio_driver.motor_direction(elevator_state.direction as u8);
                }
            }
        }
        elevator::Behaviour::DoorOpen => {
            // if requests::should_clear_immediately(&elevator_state, request_floor, button) {
            //     elevator_state.start_door_timer();
            // } else {
            //     elevator_state.set_request(request_floor, button, true);
            // }
        }
        elevator::Behaviour::Moving => {
            // elevator_state.set_request(request_floor, button, true);
        }
    }

    set_all_lights(elevio_driver, elevator_state);

    // debug!("New state: "\);
    // elevator_state.print();
}

pub fn on_arrival(
    elevio_driver: &e::Elevator,
    mut elevator_state: &mut elevator::State,
    new_floor: u8,
) {
    // debug!("Arrived at floor {}", new_floor);
    // elevator_state.print();

    elevator_state.set_floor(new_floor);
    elevio_driver.floor_indicator(new_floor);

    match elevator_state.behaviour {
        elevator::Behaviour::Moving => {
            if requests::should_stop(&elevator_state) {
                elevio_driver.motor_direction(e::DIRN_STOP);
                elevio_driver.door_light(true);
                elevator_state = requests::clear_at_current_floor(elevator_state);
                elevator_state.start_door_timer();
                set_all_lights(elevio_driver, elevator_state);
                elevator_state.behaviour = elevator::Behaviour::DoorOpen;
            }
        }
        _ => {}
    }

    // debug!("New state: "\);
    // elevator_state.print();
}

pub fn on_door_timeout(elevio_driver: &e::Elevator, mut elevator_state: &mut elevator::State) {
    // debug!("Door timeout");
    // elevator_state.print();

    match elevator_state.behaviour {
        elevator::Behaviour::DoorOpen => {
            let direction_behaviour_pair = requests::choose_direction(&elevator_state);
            elevator_state.direction = direction_behaviour_pair.0;
            elevator_state.behaviour = direction_behaviour_pair.1;

            match elevator_state.behaviour {
                elevator::Behaviour::Moving | elevator::Behaviour::Idle => {
                    elevio_driver.door_light(false);
                    elevio_driver.motor_direction(elevator_state.direction as u8);
                }
                elevator::Behaviour::DoorOpen => {
                    elevator_state.start_door_timer();
                    elevator_state = requests::clear_at_current_floor(elevator_state); // THIS NEEDS TO BE BROADCAST AS WELL
                    set_all_lights(elevio_driver, elevator_state);
                }
            }
        }
        _ => {}
    }

    // debug!("New state: "\);
    // elevator_state.print();
}
